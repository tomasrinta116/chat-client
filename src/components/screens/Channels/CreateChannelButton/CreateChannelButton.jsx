import React from "react";
import Fab from "@material-ui/core/Fab";
import AddIcon from "@material-ui/icons/Add";
import TextField from "@material-ui/core/TextField";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import Button from "@material-ui/core/Button";
import PropTypes from 'prop-types';

class CreateChannelButton extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
      channelName: ""
    };
  }

  handleClick = () => {
    this.setState({ isOpen: true });
  };

  handleClose = () => {
    this.setState({ isOpen: false });
  };

  handleChange = event => {
    this.setState({
      channelName: event.target.value
    });
  };

  createChannel = async () => {
    const { channelName } = this.state;
    const { sb, enterChannel } = this.props;
    if (!channelName) return;
    let newChannelURL = await (() => {
      const userID = sb.currentUser.userId;
      return new Promise(resolve => {
        // Array adds the operatorID's to the channel to provide admin privs.
        sb.OpenChannel.createChannel(
          channelName,
          null,
          null,
          ["admin", userID],
          (channel, error) => {
            if (error) return console.log(error);
            resolve(channel.url);
          }
        );
      });
    })();
    await enterChannel(newChannelURL);
    this.props.history.push(`/chat/${newChannelURL}`);
  };

  render() {
    return (
      <React.Fragment>
        <Fab
          data-testid="create-channel-button"
          color="primary"
          onClick={this.handleClick}
        >
          <AddIcon />
        </Fab>
        <Dialog open={this.state.isOpen} onClose={this.handleClose}>
          <DialogTitle id="form-dialog-title">Create Channel</DialogTitle>
          <DialogContent>
            <TextField
              autoFocus
              margin="dense"
              label="Channel Name"
              fullWidth
              onChange={this.handleChange}
            />
          </DialogContent>
          <DialogActions>
            <Button onClick={this.handleClose} color="primary">
              Cancel
            </Button>
            <Button id="createChannel" onClick={this.createChannel} color="primary">
              Create
            </Button>
          </DialogActions>
        </Dialog>
      </React.Fragment>
    );
  }
}

CreateChannelButton.propTypes = {
  sb: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  enterChannel: PropTypes.func.isRequired,
}

export default CreateChannelButton;
