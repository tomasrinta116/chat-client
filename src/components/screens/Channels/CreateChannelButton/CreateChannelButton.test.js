import React from "react";
import { shallow } from "enzyme";
import CreateChannelButton from "./CreateChannelButton";
import sinon from "sinon";

describe("<CreateChannelButton />", () => {
  let wrapper;

  const props = {
    sb: {},
    history: {},
    enterChannel: jest.fn(),
  }

  beforeEach(() => {
    wrapper = shallow(<CreateChannelButton {...props} />);
  });

  test("it renders only the button when isOpen=false", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("it renders the Dialog when isOpen=true/ + button has been clicked", () => {
    wrapper.setState({
      isOpen: true
    });
    expect(wrapper).toMatchSnapshot();
  });

  test("entering a channel name into the dialog updates state->channelName", () => {
    wrapper.setState({
      isOpen: true
    });
    const inputField = wrapper.find({ label: "Channel Name" });
    expect(inputField).toHaveLength(1);
    inputField.simulate("change", { target: { value: "test channel" } });
    expect(wrapper.state().channelName).toEqual("test channel");
  });

  test("clicking the create channel button triggers the createChannel function", () => {
    const spy = sinon.spy(wrapper.instance(), "createChannel");
    wrapper.setState({
      isOpen: true
    });
    const createButton = wrapper.find({ id: "createChannel" });
    expect(createButton).toHaveLength(1);
    createButton.simulate("click");
    //console.log(spy.returned(null));
    expect(spy.called).toBe(true);
  });

});
