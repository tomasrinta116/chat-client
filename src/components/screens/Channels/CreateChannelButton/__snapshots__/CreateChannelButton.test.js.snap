// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateChannelButton /> it renders only the button when isOpen=false 1`] = `
<Fragment>
  <WithStyles(Fab)
    color="primary"
    data-testid="create-channel-button"
    onClick={[Function]}
  >
    <pure(AddIcon) />
  </WithStyles(Fab)>
  <WithStyles(Dialog)
    onClose={[Function]}
    open={false}
  >
    <WithStyles(DialogTitle)
      id="form-dialog-title"
    >
      Create Channel
    </WithStyles(DialogTitle)>
    <WithStyles(DialogContent)>
      <TextField
        autoFocus={true}
        fullWidth={true}
        label="Channel Name"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        variant="standard"
      />
    </WithStyles(DialogContent)>
    <WithStyles(DialogActions)>
      <WithStyles(Button)
        color="primary"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(Button)>
      <WithStyles(Button)
        color="primary"
        id="createChannel"
        onClick={[Function]}
      >
        Create
      </WithStyles(Button)>
    </WithStyles(DialogActions)>
  </WithStyles(Dialog)>
</Fragment>
`;

exports[`<CreateChannelButton /> it renders the Dialog when isOpen=true/ + button has been clicked 1`] = `
<Fragment>
  <WithStyles(Fab)
    color="primary"
    data-testid="create-channel-button"
    onClick={[Function]}
  >
    <pure(AddIcon) />
  </WithStyles(Fab)>
  <WithStyles(Dialog)
    onClose={[Function]}
    open={true}
  >
    <WithStyles(DialogTitle)
      id="form-dialog-title"
    >
      Create Channel
    </WithStyles(DialogTitle)>
    <WithStyles(DialogContent)>
      <TextField
        autoFocus={true}
        fullWidth={true}
        label="Channel Name"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        variant="standard"
      />
    </WithStyles(DialogContent)>
    <WithStyles(DialogActions)>
      <WithStyles(Button)
        color="primary"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(Button)>
      <WithStyles(Button)
        color="primary"
        id="createChannel"
        onClick={[Function]}
      >
        Create
      </WithStyles(Button)>
    </WithStyles(DialogActions)>
  </WithStyles(Dialog)>
</Fragment>
`;
